File: 1_to_100/1/solution.py

def two_sum_brute_force(nums, target):
    n = len(nums)
    for i in range(n):  # O(n)
        for j in range(i+1,n): #O(n)
            if nums[j]+nums[i] == target:
                return [i, j]

def two_sum_sortings(nums,target):
    # create number and index pairs for number list
    nums_index = [(v, index) for index, v in enumerate(nums)] # o(n) -> space 
    # sort the list of num , index pairs 
    nums_index.sort() # o(nlog(n))

    # pointer to track the elements from left
    begin = 0
    # pointer to track the elements from right
    end = len(nums) - 1
    while begin < end: # o(n)
        # get sum of begin and end elements
        sum_of_nums = nums_index[begin][0] + nums_index[end][0]
        # return indexs if  sum mateches the target
        if sum_of_nums == target:
            return [nums_index[begin][1], nums_index[end][1]]
        # move the left point to the right by one step to increase the sum of numbers
        elif sum_of_nums < target:
            begin += 1
        # move the right point to the left by one step to decrease the sum of numbers
        else:
            end -= 1

def two_sum_hashing(nums, target):
    nums_hash ={}

    # create a hash table for each number
    for index,num in enumerate(nums):
        # try to find the number
        try:
        # append the index to the hash table
            nums_hash[num].append(index)
        # catch the KeyError and create a new key for number 
        except KeyError:
            nums_hash[num] = [index]
    for index,num in enumerate(nums):
        # get the complement of the number 
        complement = target - num
        # try to find the complement of the number in the hash table
        try:
            complement_indexes = nums_hash[complement]
            # once get the complement check the length of index list
            # to counter the case of repeating elements 
            if len(complement_indexes)>1:
                return complement_indexes[:2]
            else:
                return [index, complement_indexes[0]]
        except KeyError:
            pass



if __name__ == '__main__':
    nums = [2,7,11,15]
    target = 9
    print(f"two_sum_brute_force {two_sum_brute_force(nums, target)}")
    print(f"two sum sortings {two_sum_sortings(nums, target)}")
    print(f"two sum hashing {two_sum_hashing(nums, target)}")

File: 1_to_100/1/solution.cpp
#include <iostream>
#include <bits/stdc++.h>  
#include <vector>  
using namespace std;  


vector<int> twoSum(vector<int> &nums, int target){
        // result vector to store the result
        vector<int> result;
        // size of nums array
        int size = nums.size();
        // first pointer to track the elements of input vector
        for(int i = 0;i<size;i++){
            /*second pointer to track the elements of output vector*/
            for(int j = i+1;j<size;j++){
                // match the sum of the two elements 
                if (nums[i]+nums[j]==target){
                    // break out of loop
                    // and push indexes in result vector
                    result.push_back(i);
                    result.push_back(j);
                    break;
                }
            }
        }
     return result;   
    }
vector<int> two_sum_hashing(vector<int> & nums, int target){
        // vector to store the result
        vector<int> result(2);
        // create a map object to store array elements and their indices
        map<int, int> num_map; 
        // get the size of the array
        int n = nums.size(), temp_find = 0;
        // iterate over the array
        for(int i=0; i<n;i++){
            // check for the compliment of the element in the array
            if(num_map.count(target-num_map[i])){
                result[0] =num_map[target-num_map[i]];
                result[1] = i;
                return result;
            }
            cout <<num_map[target-nums[i]]<<endl;
            // check for the element in the map 
            if(num_map[nums[i]]==0){
                    num_map[nums[i]]=i;
                    
            }
            
        }
        result[0] = 0;
        result[1] = 0;
        return result;
}
int main(){
    vector<int> nums;
    nums.push_back(2);
    nums.push_back(7);
    nums.push_back(11);
    nums.push_back(15);
    int target = 9;
    // nums.push_back()
    vector<int> result = two_sum_hashing(nums, target);
    for(int i = 0; i < 2; i++){
           cout <<result[i] << endl;
    }


}